### PR TITLE
Add timers for NEST 2.14.1

### PR DESCRIPTION
**Pull request by jarsi**
This PR adds timer functionality to NEST 2.14.1. The usage and functionality is similar to what is implemented in PR #1778 and addresses the issues raised in #1471.

These timers are implemented:
* `time_gather_spike_data`
* `time_simulate` (basic timer, always enabled)
    * nested in `time_simulate:`
        * `time_update`
        * `time_gather_spike_data`
            * nested in `time_gather_spike_data:`
                * `time_collocate_spike_data`
                * `time_communicate_spike_data`
                * `time_deliver_spike_data`

In contrast to PR #1778, these timers are not implemented due to the infrastructure change:
* `time_communicate_prepare` (basic timer, always enabled)
    * nested in `time_communicate_prepare:`
        * `time_gather_target_data`
            * nested in `time_gather_target_data:`
                * `time_communicate_target_data`

I am not too familiar with the construction phases. Furthermore the order of collocate, communicate and delivery is different and I am not sure how this influences the `time_gather_spike_data`. So besides the normal review process, I'd ask specifically for input for these parts:

* `time_construction_create` (basic timer, always enabled)
* `time_construction_connect` (basic timer, always enabled)
* `time_gather_spike_data`

### CODE REVIEW

**Comment by jarsi** [comment]
Could one of you, @stinebuu or @hakonsbm, look at the use of `time_construction_create` and `time_construction_connect`? I could imagine these timers need some correction.

@jakobj, could you check whether `time_gather_spike_data` is implemented correctly?

File: nestkernel/nest_names.cpp [hunk 1]
```diff
@@ -466,8 +466,18 @@ const Name theta_eq( "theta_eq" );
...
+const Name time_update( "time_update" );
+#endif
+const Name time_construction_connect( "time_construction_connect" );
+const Name time_construction_create( "time_construction_create" );
 const Name tics_per_step( "tics_per_step" );
 const Name time_in_steps( "time_in_steps" );
+const Name time_simulate( "time_simulate" );
 const Name time( "time" );
 const Name time_collocate( "time_collocate" );
 const Name time_communicate( "time_communicate" );
```
Comment (hakonsbm) [inline nestkernel/nest_names.cpp:483]: Should these be removed too?

File: nestkernel/simulation_manager.cpp [hunk 2]
```diff
@@ -71,6 +71,18 @@ nest::SimulationManager::initialize()
   simulated_ = false;
   exit_on_user_signal_ = false;
   inconsistent_state_ = false;
+  // reset_timers_for_dynamics();
```
Comment (hakonsbm) [inline nestkernel/simulation_manager.cpp:74]:
```suggestion
```

File: nestkernel/connection_manager.h [hunk 3]
```diff
@@ -294,6 +295,10 @@ class ConnectionManager : public ManagerInterface
    */
   DelayChecker& get_delay_checker();
 
+  // public stop watch for benchmarking purposes
+  // start and stop in high-level connect functions in nestmodule.cpp and nest.cpp
+  Stopwatch sw_construction_connect;
```
Comment (hakonsbm) [inline nestkernel/connection_manager.h:300]: This should have a doxygen documentation comment, e.g. `Stopwatch measuring connection time`.

File: nestkernel/event_delivery_manager.h [hunk 4]
```diff
@@ -320,23 +324,18 @@ class EventDeliveryManager : public ManagerInterface
...
    * Number of generated spike events (both off- and on-grid) during the last
    * call to simulate.
    */
   unsigned long local_spike_counter_;
+
+#ifdef TIMER_DETAILED
+  // private stop watches for benchmarking purposes
+  // (intended for internal core developers, not for use in the public API)
+  Stopwatch sw_collocate_spike_data_;
+  Stopwatch sw_communicate_spike_data_;
```
Comment (hakonsbm) [inline nestkernel/event_delivery_manager.h:337]: These should also have doxygen comments.

File: nestkernel/node_manager.h [hunk 5]
```diff
@@ -308,6 +309,9 @@ class NodeManager : public ManagerInterface
   //! Network size when nodes_vec_ was last updated
   index nodes_vec_network_size_;
   size_t num_active_nodes_; //!< number of nodes created by prepare_nodes
+
+  // private stop watch for benchmarking purposes
+  Stopwatch sw_construction_create_;
```
Comment (hakonsbm) [inline nestkernel/node_manager.h:314]: Add doxygen comment.

**Review by hakonsbm — CHANGES_REQUESTED** [review]
The `time_construction_create` timer looks fine.

For `time_construction_connect` there should be timers in the `ConnectLayer` function in `topologymodule.cpp` to get connection-times for spatial connections. 

There should also be timers for `DataConnect`. Note that there are two high-level functions for `DataConnect` in `nestmodule.cpp`.

And see below for some other comments.

**Comment by jarsi** [comment]
Thanks for the review @hakonsbm. I added and changed the functionality as you suggested. Could you have a look again?

**Review by hakonsbm — APPROVED** [review]
Thanks! Looks good to me now.

**Comment by jarsi** [comment]
Thanks for you measurements. The difference in communication time is due to a now missing synchronization point. Previously this synchronization was not measured and excluded from the communication time. Now it's implicitly contained in the call to communicate.

Screenshot from 2021-06-17 10-55-28

**Review by jougs — APPROVED** [review]
LGTM!

@ackurth: could you please indicate if @jarsi's last reply settles your concerns? Thanks!

**Review by ackurth — APPROVED** [review]

**Comment by jougs** [comment]
Many thanks to everyone!

@terhorstd: let me know when you're in the mood for a release :cat:

**Pull request merged**